main.go

```diff
@@ -202,3 +215,37 @@ func wrapAuthHandler(handler http.HandlerFunc, htpasswdFile string, host string)
 	authenticator := auth.NewBasicAuthenticator(host, secrets)
 	return auth.JustCheck(authenticator, handler)
 }
+
+func writePidFile(c *config.Config, accessLogger cache.Logger) error {
+	// create a "pid" directory under cache directory and clean up inactive pid file
+	pidPath := filepath.Join(c.Dir, "pid")
+	err := os.MkdirAll(pidPath, os.FileMode(0744))
+	if err != nil {
+		log.Fatal(err)
+		return err
+	}
+	err = filepath.Walk(pidPath, func(name string, fileInfo os.FileInfo, err error) error {
+		if !fileInfo.IsDir() {
+			pid, _ := strconv.Atoi(fileInfo.Name())
+			bazelRemoteProcess, err := os.FindProcess(pid)
+			if err != nil {
+				accessLogger.Printf("Error to find pid: %d", bazelRemoteProcess)
+			} else {
+				err = bazelRemoteProcess.Signal(syscall.Signal(0))
+				if err != nil && strings.Contains(err.Error(), "process already finished") {
+					accessLogger.Printf("Removing pid file: %s", fileInfo.Name())
+					err = os.Remove(filepath.Join(c.Dir, "pid", fileInfo.Name()))
```

Comment:
Where is the `pid` file removed if the server exits by itself? For example because of an idle timeout? Or Ctrl-C'd? Or killed from a different terminal?

Reply:
The pid file is written at bazel-remote start up and is removed by another instance start up. The purpose of a pid file is just to let bazel-remote gain knowledge of other PIDs

Reply:
What if there isn't another instance and the only `bazel-remote` instance just exits due to an `idle_timeout`? Who cleans up the PID file then?

Reply:
I agree with ob that the current process should delete its pid when gracefully terminated.  Cleaning up at the next start is a way to do garbage collection, but this is not the normal case.  The normal case should be: no processing running, no pid files.

Reply:
I don't think it is good idea to rely on the file to assume if there is a pid file the process is running. The pid file is just to tell bazel-remote a process might be running. It has to verify if it is actually running, so I don't see it is necessary to hold this: "no processing running, no pid files."

Reply:
No one has to RELY on the file to assume if there is a pid file the process is running. But it is a good indication. It is a common practice of using pid files